GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs

```diff
@@ -1554,6 +1554,14 @@ private string GetNewProjectedShaForPlaceholder(string path)
                             {
                                 this.placeholderDatabase.AddPartialFolder(childRelativePath);
                             }
+                            else if (result.Result == FSResult.IOError)
+                            {
+                                if (this.context.FileSystem.DirectoryExists(Path.Combine(this.context.Enlistment.WorkingDirectoryRoot, childRelativePath)))
+                                {
+                                    this.ReExpandFolder(childRelativePath, existingPlaceholders);

```

Comment:
Offline you mentioned:

> since I changed to no longer call checkout -f and instead use the postindexchanged there were folders that were on disk but not in the placeholder list and would get an error when trying to write the placeholder directory

But I'm not quite following the scenario.  How/why are there already folders on disk, and how is it related to no longer calling `checkout -f`?  Are these folders full folders?

I was also wondering if there's any chance that the folders we're re-expanding here (or their children) are already in the placeholder list as expanded?

Reply:
The test that manifested this issue was `DeleteFileAndCommitThenChangingSparseFoldersShouldKeepFolderAndFile` which:
1. Deletes `GVFS\GVFS.Tests\packages.config` and commits that change so the path for that file is in the modified paths.  
2. It then adds `GVFS\GVFS` as a sparse folder so the file that was deleted is outside the sparse set and verifies that the file is still deleted.  
3. Then it checks out `HEAD~1` and verifies that the file comes back which it does because it is in the modified paths.  This checkout causes the `GVFS\GVFS.Tests` folder to be removed from the placeholders because it is not part of the projection since only `GVFS\GVFS` is being projected.
4. In the teardown all the sparse folders are removed which should return the repo to being fully projected, but when reexpanding the `GVFS` folder it tries to make a placeholder directory for `GVFS.Tests` but it already exists.

I think this would have been failing with the `checkout -f` as well because the `GVFS.Tests` directory would have still been on disk and the `WritePlaceholderDirectory` call would have failed.  So this was a bug in the original sparse code but not caught because the tests were only validating the top level folders and `DeleteFileAndCommitThenChangingSparseFoldersShouldKeepFolderAndFile` was the last test to run so there were not any subsequent tests that might have caught it.